cpp/examples/arrow/rapidjson_row_converter.cc

```diff
@@ -0,0 +1,590 @@
+    rows_.reserve(num_rows);
+
+    for (int64_t i = 0; i < num_rows; ++i) {
+      rows_.push_back(rapidjson::Document());
+      rows_[i].SetObject();
+    }
+  }
+
+  /// \brief Set which field to convert.
+  void SetField(std::shared_ptr<arrow::Field> field) { field_ = std::move(field); }
+
+  /// \brief Retrieve converted rows from builder.
+  std::vector<rapidjson::Document> Rows() && { return std::move(rows_); }
+
+  // Default implementation
+  arrow::Status Visit(const arrow::Array& array) {
+    return arrow::Status::NotImplemented(
+        "Can not convert to json document for array of type ", array.type()->ToString());
+  }
+
+  // Handles booleans, integers, floats
+  template <typename ArrayType, typename DataClass = typename ArrayType::TypeClass>
+  arrow::enable_if_primitive_ctype<DataClass, arrow::Status> Visit(
+      const ArrayType& array) {
+    for (int64_t i = 0; i < array.length(); ++i) {
+      if (!array.IsNull(i)) {
+        rapidjson::Value str_key(field_->name().c_str(), rows_[i].GetAllocator());
+        rows_[i].AddMember(str_key, array.Value(i), rows_[i].GetAllocator());
+      }
+    }
+    return arrow::Status::OK();
+  }
+
+  arrow::Status Visit(const arrow::StringArray& array) {
+    for (int64_t i = 0; i < array.length(); ++i) {
+      if (!array.IsNull(i)) {
+        rapidjson::Value str_key(field_->name().c_str(), rows_[i].GetAllocator());
+        // StringArray.Value returns a string view
+        auto value_view = array.Value(i);
+        rapidjson::Value value;
+        value.SetString(value_view.data(),
+                        static_cast<rapidjson::SizeType>(value_view.size()),
+                        rows_[i].GetAllocator());
+        rows_[i].AddMember(str_key, value, rows_[i].GetAllocator());
+      }
+    }
+    return arrow::Status::OK();
+  }
+
+  arrow::Status Visit(const arrow::StructArray& array) {
+    const arrow::StructType* type = array.struct_type();
+
+    RowBatchBuilder child_builder(rows_.size());
+    for (int i = 0; i < type->num_fields(); ++i) {
+      std::shared_ptr<arrow::Field> child_field = type->field(i);
+      child_builder.SetField(std::move(child_field));
+      ARROW_RETURN_NOT_OK(arrow::VisitArrayInline(*array.field(i).get(), &child_builder));
+    }
+    std::vector<rapidjson::Document> rows = std::move(child_builder).Rows();
+
+    for (int64_t i = 0; i < array.length(); ++i) {
+      if (!array.IsNull(i)) {
+        rapidjson::Value str_key(field_->name().c_str(), rows_[i].GetAllocator());
+        // Must copy value to new allocator
+        rapidjson::Value row_val;
+        row_val.CopyFrom(rows[i], rows_[i].GetAllocator());
+        rows_[i].AddMember(str_key, row_val, rows_[i].GetAllocator());
+      }
+    }
+    return arrow::Status::OK();
+  }
+
+  arrow::Status Visit(const arrow::ListArray& array) {
+    // First create rows from values
+    std::shared_ptr<arrow::Array> values = array.values();
```

Comment:
While minor, I wonder why `BaseListArray::values` doesn't return `const Array&` or at least `const std::shared_ptr<Array>&`…not that it should be a big deal here.